### PR TITLE
GPG signature verification overhaul. GPG2, HKPS, gpgv2.

### LINKED ISSUE #1199: openconnect: GPG key retrieval failure, localhost or remote (existing srv)
### Expected behavior:
just should install.

### Actual Behavior:
using streisand install  "existing server (8)" or "localhost (7)", on a pristine ubu16.4, both fails with see below.
sparing out openconnect, doesn't help. see below.

### Steps to Reproduce:
1. ./streisand
install all
no domain/email bcs it's a local kvm (testing)

### Ansible Information
* Ansible version: 2.4.3.0
* Ansible system: Linux
* Host OS: Ubuntu
* Host OS version:  16.04
* Python interpreter: /usr/bin/python
* Python version: 2.7.12

### Streisand Information
* Streisand Git revision: 43f474b79a18a71bd670bb0e1f086f172a72a481
* Streisand Git clone has untracked changes: no
* Genesis role: localhost
* Custom SSH key: False

### Enabled Roles
* Shadowsocks enabled:  True
* Wireguard enabled: True
* OpenVPN enabled: True
* stunnel enabled: True
* Tor enabled: True
* Openconnect enabled: True
* TinyProxy enabled: True
* SSH forward user enabled: True
* L2TP enabled: True
* Configured number of VPN clients: 5

```
TASK [openconnect : Download the ocserv source code and signature file] ********************************************************************************************************************************************************************

TASK [download-and-verify : Nikos Mavrogiannopoulos signs the OpenConnect downloads. Import the correct GPG key.] **************************************************************************************************************************
ok: [192.168.33.153]

TASK [download-and-verify : Register the GPG Key fingerprint output] ***********************************************************************************************************************************************************************
ok: [192.168.33.153]

TASK [download-and-verify : Make sure the retrieved fingerprint perfectly matches the expected fingerprint] ********************************************************************************************************************************
fatal: [192.168.33.153]: FAILED! => {
    "assertion": "project_expected_fingerprint in retrieved_fingerprint.stdout", 
    "changed": false, 
    "evaluated_to": false
}
```
### Ansible Information
* Ansible version: 2.4.3.0
* Ansible system: Linux
* Host OS: Debian
* Host OS version:  9.3
* Python interpreter: python
* Python version: 2.7.13

### Streisand Information
* Streisand Git revision: 43f474b79a18a71bd670bb0e1f086f172a72a481
* Streisand Git clone has untracked changes: no
* Genesis role: existing-server
* Custom SSH key: False

### Enabled Roles
* Shadowsocks enabled:  True
* Wireguard enabled: True
* OpenVPN enabled: True
* stunnel enabled: True
* Tor enabled: True
* Openconnect enabled: False
* TinyProxy enabled: True
* SSH forward user enabled: True
* L2TP enabled: True
* Configured number of VPN clients: 5

```
TASK [l2tp-ipsec : Retrieve the Libreswan source code] *************************************************************************************************************************************************************************************

TASK [download-and-verify : The Libreswan team signs the The Libreswan Project downloads. Import the correct GPG key.] *********************************************************************************************************************
ok: [192.168.33.153]

TASK [download-and-verify : Register the GPG Key fingerprint output] ***********************************************************************************************************************************************************************
ok: [192.168.33.153]

TASK [download-and-verify : Make sure the retrieved fingerprint perfectly matches the expected fingerprint] ********************************************************************************************************************************
fatal: [192.168.33.153]: FAILED! => {
    "assertion": "project_expected_fingerprint in retrieved_fingerprint.stdout", 
    "changed": false, 
    "evaluated_to": false
}
```

### PR DESCRIPTION
**Pull request by cpu**
This commit revamps the Streisand GPG signature validation and GPG key management.

Prior to this commit we had a mix of "baked in" GPG keys and places where the `download-and-verify` playbook was expected to fetch a key from a keyserver. After this commit a central `gpg` playbook is used to set up a keyring with known good keys that can be kept up to date. Overall this approach should be more secure and hopefully a lot less flaky for end users experiencing flaky keyserver errors.

Part of this revamp switches to using `gpgv2` to do the actual signature validation. This is a tool [specifically designed for this job](https://www.gnupg.org/documentation/manuals/gnupg/gpgv.html) and contains a lot less functionality than the base `gpg2` binary. One of the handy things about `gpgv2` is it assumes the keys in the provided keyring are ultimately trusted. In contrast, when using `gpg2` to verify signatures you have to either do the whole `gpg2 edit <keyid>`, `trust`, `5`, `save` dance for each key (interactive, annoying), or do that once and then use `--export-ownertrust` and `--import-ownertrust` (workable, but more code!).

In preparation for being able to sign our own content & provide options for secure communication/filesharing between the Streisand maintainers and Streisand users this revamp also adds all of the active maintainer's GPG keys to the Streisand keyring. _To ward off any potential suspicion: The Streisand keyring is separate from the system keyring and can not be used to verify APT packages and there is no connection between these GPG public keys and system/ssh access._

The old code had a few things I think were downsides:
* It had to rename signature files to expected extensions
* It globbed the inputs to the `gpg` binary in a way that meant one project's `download-and-verify` run was validating other project's signatures when they all downloaded to the same dir (e.g. `/usr/local/src`)
* It used the full-fledged `gpg` binary to validate signatures.
* Some software didn't have a baked in key in the repo so any keyserver search failures meant a failure to validate a signature and a provisioning error
* The key's that were imported from the repo or from keyserver search were not marked trusted, so all of the signature validations had warnings about key trust.
* The key ID matching/good signature matching checked the verify output in a way that could potentially by spoofed by attacker controlled content in the GPG keys (e.g. the UID strings)
* The default key search timeout was pretty low, and key search was done over plaintext HTTP on a non-standard port. This seemed to be flaky in practice based on user bug reports.

New code highlights:

* Uses GPG2 instead of GPG. Modern!
* Builds a standalone keyring for Streisand in `.gnupg/streisand`. Clean!
* Sets up GPG2/`dirmngr` to fetch keys over HTTPS with HKPS to the [HKPS Keyservers.net pool](https://sks-keyservers.net/overview-of-pools.php#pool_hkps). Privacy++!
* Increases the key fetch timeouts. Reliability++?
* Sets up a cronjob to refresh the Streisand keyring periodically. Up-to-date!
* Pre-initializes keyring with known-good keys from the repo. Resilience++
* Signature verification is done with `gpgv2` instead of `gpg`. Security++
* `gpgv2` verification uses exact filenames/paths instead of globbing. Security++
* Key ID enforcement is strict, with no chance of matching a key ID from a UID field. Security++
* Duplicated download/verification code from the `tor-bridge` playbook was replaced by `download-and-verify` usage. Tidy!
* Maybe some other stuff? I forget!

The following GPG keys are imported to the Streisand keyring:
```
# OpenVPN release signing key
- 2F2B01E7.security@openvpn.net.asc
# Openconnect
- 7F343FA7.nmav@redhat.com.asc
- 96865171.nmav@gnutls.org.asc
# Tor browser release signing key
- 93298290.torbrowser@torproject.org.asc
# PuTTY release signing key
- B43434E4.putty@projects.tartarus.org.asc
# Stunnel release signing key
- DD3AAAA3.Michal.Trojnara@stunnel.org.asc
# LibreSwan release signing key
- B30FC6F9.team@libreswan.org.asc
# Streisand maintainer - Github @cpu
- 2D8330C2.daniel@binaryparadox.net.asc
# Streisand maintainer - Github @jlund
- CDF6583E.josh@joshlund.com.asc
# Streisand maintainer - Github @nopdotcom
- F67DA905.nop@nop.com.asc
# Streisand maintainer - Github @CorbanR
- 3F7F585B.corban@raunco.co.asc
```

I've left out @alimakki's key temporarily because it seems to mysteriously break in CI but works when testing real hosts :confused: :question: 

For each piece of respective software we pin the specific key ID from the keyring for verifying releases:
* OpenVPN -> `AF131CAE`
* OpenConnect -> `9013B842`
* PuTTY -> `B43434E4`
* Stunnel -> `D416E014`
* Libreswan -> `B30FC6F9`

It's safe to refer to these keys by short ID prefixes strictly because we are using a prepopulated and separated GPG keyring and not searching keyservers for the initial import.

Resolves #1199, #1227, #963

### CODE REVIEW

**Comment by nopdotcom** [comment]
I'm in favor of gpg2. Would `gpgv` work the same way as `gpgv2`? I can think of only one reason why `gpgv` would make any sense, which is bare Ubuntu. That's not a particularly good reason given how much software we depend on.

```
# dpkg -l gpgv gpgv2
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                       Version            Architecture       Description
+++-==========================-==================-==================-==========================================================
ii  gpgv                       1.4.20-1ubuntu3.1  amd64              GNU privacy guard - signature verification tool
un  gpgv2                      <none>             <none>             (no description available)
```

File: playbooks/roles/download-and-verify/tasks/main.yml [hunk 1]
```diff
@@ -1,41 +1,50 @@
...
-  shell: for file in *.asc; do gpg --verify $file; done
-  args:
-    chdir: "{{ project_download_location }}"
+- name: "Verify the {{ project_name }} download signatures with the Streisand GPG keyring"
+  command: "gpgv2 --keyring {{ streisand_gpg_keyring }} {{ project_download_location }}/{{ item.sig }} {{ project_download_location }}/{{ item.file }}"
   register: gpg_verification_results
-  changed_when: False
+  with_items: "{{ project_download_files }}"
+
+- name: "Verify the {{ project_name }} download signature checks all passed"
```
Comment (nopdotcom) [inline playbooks/roles/download-and-verify/tasks/main.yml]: This seems a little spammy compared to other stuff. 

```
TASK [download-and-verify : Verify the OpenConnect download signature checks all passed] ******************************************
ok: [localhost] => (item={'_ansible_parsed': True, 'stderr_lines': [u'gpgv: Signature made Sun 07 Jan 2018 03:36:54 PM UTC using RSA key ID 9013B842', u'gpgv: Good signature from "Nikos Mavrogiannopoulos <nmav@gnutls.org>"', u'gpgv:                 aka "Nikos Mavrogiannopoulos <nmav@hushmail.com>"', u'gpgv:                 aka "Nikos Mavrogiannopoulos <n.mavrogiannopoulos@gmail.com>"'], '_ansible_item_result': True, u'end': u'2018-03-26 23:35:25.680773', '_ansible_no_log': False, u'stdout': u'', u'cmd': [u'gpgv2', u'--keyring', u'/root/.gnupg/streisand/pubring.gpg', u'/usr/local/src//ocserv-0.11.10.tar.xz.sig', u'/usr/local/src//ocserv-0.11.10.tar.xz'], u'rc': 0, 'item': {u'sig': u'ocserv-0.11.10.tar.xz.sig', u'file': u'ocserv-0.11.10.tar.xz'}, u'delta': u'0:00:00.009391', u'stderr': u'gpgv: Signature made Sun 07 Jan 2018 03:36:54 PM UTC using RSA key ID 9013B842\ngpgv: Good signature from "Nikos Mavrogiannopoulos <nmav@gnutls.org>"\ngpgv:                 aka "Nikos Mavrogiannopoulos <nmav@hushmail.com>"\ngpgv:                 aka "Nikos Mavrogiannopoulos <n.mavrogiannopoulos@gmail.com>"', u'changed': True, u'invocation': {u'module_args': {u'creates': None, u'executable': None, u'_uses_shell': False, u'_raw_params': u'gpgv2 --keyring /root/.gnupg/streisand/pubring.gpg /usr/local/src//ocserv-0.11.10.tar.xz.sig /usr/local/src//ocserv-0.11.10.tar.xz', u'removes': None, u'warn': True, u'chdir': None, u'stdin': None}}, 'stdout_lines': [], u'start': u'2018-03-26 23:35:25.671382', '_ansible_ignore_errors': None, 'failed': False}) => {
    "changed": false,
    "item": {
        "changed": true,
        "cmd": [
            "gpgv2",
            "--keyring",
            "/root/.gnupg/streisand/pubring.gpg",
            "/usr/local/src//ocserv-0.11.10.tar.xz.sig",
            "/usr/local/src//ocserv-0.11.10.tar.xz"
        ],
        "delta": "0:00:00.009391",
        "end": "2018-03-26 23:35:25.680773",
        "failed": false,
        "invocation": {
            "module_args": {
                "_raw_params": "gpgv2 --keyring /root/.gnupg/streisand/pubring.gpg /usr/local/src//ocserv-0.11.10.tar.xz.sig /usr/local/src//ocserv-0.11.10.tar.xz",
                "_uses_shell": false,
                "chdir": null,
                "creates": null,
                "executable": null,
                "removes": null,
                "stdin": null,
                "warn": true
            }
        },
        "item": {
            "file": "ocserv-0.11.10.tar.xz",
            "sig": "ocserv-0.11.10.tar.xz.sig"
        },
        "rc": 0,
        "start": "2018-03-26 23:35:25.671382",
        "stderr": "gpgv: Signature made Sun 07 Jan 2018 03:36:54 PM UTC using RSA key ID 9013B842\ngpgv: Good signature from \"Nikos Mavrogiannopoulos <nmav@gnutls.org>\"\ngpgv:                 aka \"Nikos Mavrogiannopoulos <nmav@hushmail.com>\"\ngpgv:                 aka \"Nikos Mavrogiannopoulos <n.mavrogiannopoulos@gmail.com>\"",
        "stderr_lines": [
            "gpgv: Signature made Sun 07 Jan 2018 03:36:54 PM UTC using RSA key ID 9013B842",
            "gpgv: Good signature from \"Nikos Mavrogiannopoulos <nmav@gnutls.org>\"",
            "gpgv:                 aka \"Nikos Mavrogiannopoulos <nmav@hushmail.com>\"",
            "gpgv:                 aka \"Nikos Mavrogiannopoulos <n.mavrogiannopoulos@gmail.com>\""
        ],
        "stdout": "",
        "stdout_lines": []
    },
    "msg": "All assertions passed"
}

TASK [download-and-verify : Verify the OpenConnect download signatures were from the correct keys] ****************
```

File: playbooks/roles/download-and-verify/tasks/main.yml [hunk 2]
```diff
@@ -1,41 +1,50 @@
...
+  assert:
+    that:
+      - "not item.failed"
+      - "item.rc == 0"
+    msg: "Verifying {{ item.item.file }} GPG signature failed!!"
+  with_items: "{{ gpg_verification_results.results }}"
 
-- name: Make sure the {{ project_name }} files all passed GPG signature verification
-  assert: { that: "'BAD signature' not in gpg_verification_results.stderr" }
+- name: "Verify the {{ project_name }} download signatures were from the correct keys"
```
Comment (nopdotcom) [inline playbooks/roles/download-and-verify/tasks/main.yml]: This is kinda long too. Your call on the matter of taste; personally, I'd want to keep the diagnostics in until everything else is done....

File: playbooks/roles/download-and-verify/tasks/main.yml (hunk 1 above)
Comment (cpu) [reply playbooks/roles/download-and-verify/tasks/main.yml]: Agreed! I meant to try and find a way to silence this super verbose `assert` output and then forgot about it. I'll take that as an item to clean-up.

**Comment by cpu** [comment]
> Would gpgv work the same way as gpgv2? 

I'm not sure I understand the question. Will it work the same way in what capacity? It verifies GPG signatures the same way, the keyring handling is different and basically everything but signature verification is removed.

> I can think of only one reason why gpgv would make any sense, which is bare Ubuntu. That's not a particularly good reason given how much software we depend on.

`playbooks/roles/gpg/tasks/main.yml` ensures its installed alongside `gnupg2` (which also isn't on most bare Ubuntu 16.04 images).

**Comment by nopdotcom** [comment]
Got it; I hadn't looked into the keyring differences between gpg and gpg2 in a long time. That's what I get for using frontends.

File: playbooks/roles/download-and-verify/tasks/main.yml (hunk 1 above)
Comment (cpu) [reply playbooks/roles/download-and-verify/tasks/main.yml]: This turns out to be a real pain in the butt. There's an open upstream Ansible issue (https://github.com/ansible/ansible/issues/27124) about the verbosity of the `assert` task but unfortunately it seems like its stalled somewhere around what I'd consider WONTFIX.

I've [bumped that issue](https://github.com/ansible/ansible/issues/27124#issuecomment-376523736) to try and make the case that this is really unfortunate behaviour that should be fixed.

Comment (cpu) [reply playbooks/roles/download-and-verify/tasks/main.yml]: It's unlikely I'll be able to come up with a clever solution to this in the next day or so and this PR fixes a bug multiple users have reported. In the interest of moving it forwards I added a `no_log` to each of the asserts and the output is much better now, e.g.:

```
TASK [download-and-verify : Verify the The Libreswan Project download signature checks all passed] ***
ok: [localhost] => (item=(censored due to no_log)) => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result"}
TASK [download-and-verify : Verify the The Libreswan Project download signatures were from the correct keys] ***
ok: [localhost] => (item=(censored due to no_log)) => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result"}
```

We'll have to disable `no_log` if the asserts fail to debug further.

@nopdotcom Does this resolve your review feedback?

Comment (nopdotcom) [reply playbooks/roles/download-and-verify/tasks/main.yml]: Yes, and you put way more effort into it than I thought! Thanks.

**Comment by cpu** [comment]
I haven't been able to drum up any other reviewers. I feel confident in the shape of this PR. It's been running nicely in CI, I've had good results on test instances and so has @nopdotcom. I think it's a pretty straight win over the existing code. I'm going to hit the merge button and keep an eye out for issues cropping out.

Thanks all!

**Pull request merged**